src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+  }
+
+  @Override
+  protected boolean preHandle(ServletRequest request, ServletResponse response) throws Exception {
+
+    try {
+        logger.debug("preHandle in TeamProjectBasedAuthorizingFilter == '{}'", this.authorizationMode);
+        if (this.authorizationMode.equals("teamproject") && SecurityUtils.getSubject().isAuthenticated()) {
+          // in case of "teamproject" mode, we want all roles to be reset always, and
+          // set to only the one requested/found in the request parameters (following lines below):
+          String login = this.authorizer.getCurrentUser().getLogin();
+          boolean foundValidTeamProject = extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(this, login, request, response);
+          if (!foundValidTeamProject) {
+            return false;
+          }
+        }
+
+      } catch (Exception e) {
+        WebUtils.toHttp(response).setHeader("x-auth-error", e.getMessage());
+        throw new Exception(e);
+      }
+
+    return true;
+  }
+
+  /**
+   * Tries to extract "team project" role found in the request.
+   * If necessary, assigns the newly found "team project" role to the user,
+   * after validating with a Gen3 authorization service if the user
+   * has been granted access to this team.
+   * 
+   * @param self: reference to "this". Just a workaround to allow the method to be
+   * globally synchronized by making it static.
+   * @param login: the user login
+   * @param request: the request passing through this filter
+   * @param response: response object that can be used to write error status and
+   * error message if needed.
+   * 
+   * @return: Returns true if the "team project" was found and passed validation
+   * and DB update to user roles did not fail.
+   * 
+   * @throws IOException
+   * @throws Exception
+   */
+  private static boolean extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(
+      TeamProjectBasedAuthorizingFilter self,
+      String login, ServletRequest request,
+      ServletResponse response)  throws IOException, Exception {
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
+      // check if a teamproject parameter is found in the request:
+      String teamProjectRole = self.extractTeamProjectFromRequestParameters(request);
+      Set<String> newUserRoles = new HashSet<String>();
+      Set<String> newDefaultRoles = new HashSet<String>(self.defaultRoles);
+
+      // if found, add teamproject as a role in the newUserRoles list:
+      if (teamProjectRole != null && !teamProjectRole.trim().isEmpty()) {
+        // double check if this role has really been granted to the user:
+        if (self.checkGen3Authorization(teamProjectRole, login) == false) {
+          String errorMessage = "User is not authorized to access this team project's data";
```

Comment:
Might consider a different HTTP response so as not to reveal that the Team exists e.g., 400 Bad Request

Reply:
Good point, although this 403 also returns for non-existing teams...so we've covered that.